### ISSUE TITLE
(not) referencing $& causes match failure

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11192]
Migrated from [rt.perl.org#86042](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=86042) (status was 'resolved')

Searchable as RT86042$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



t/re/reg\_unsafe\.t was added in a229a030867b4d77\, because a match fails when
run alone\, but passes when in t/re/pat\.t \(subsequently moved to pat\_advanced\.t\)

I don't think that there's a ticket for this\. I may have stumbled onto an
insight&#8203;:

$ git diff

$ cat t/re/reg\_unsafe\.t
\#\!\./perl

BEGIN \{
&nbsp;    chdir 't' if \-d 't';
&nbsp;    @&#8203;INC = '\.\./lib';

\}
print "1\.\.1\\n";

\# there is an equivalent test in t/re/pat\.t which does NOT fail
\# its not clear why it doesn't fail\, so this todo gets its own test
\# file until we can work it out\.

my $x;
\($x='abc'\)=~/\(abc\)/g;
$x='123';

print "not " if $1 ne 'abc';
print "ok 1 \# TODO safe match vars make /g slow\\n";

$ \./perl t/re/reg\_unsafe\.t
1\.\.1
not ok 1 \# TODO safe match vars make /g slow

$ git diff
<details><summary>Inline Patch</summary>

```diff
diff --git a/t/re/reg_unsafe.t b/t/re/reg_unsafe.t
index 49c7011..11a621f 100644
--- a/t/re/reg_unsafe.t
+++ b/t/re/reg_unsafe.t
@@ -11,6 +11,7 @@ print "1..1\n";
 # its not clear why it doesn't fail, so this todo gets its own test
 # file until we can work it out.

+$&;
 my $x;
 ($x='abc')=~/(abc)/g;
 $x='123';


```
</details>
$ ./perl t/re/reg_unsafe.t
1\.\.1
ok 1 \# TODO safe match vars make /g slow


ie adding $& \(and thereby setting PL\_sawampersand\) causes the bug to be "fixed"

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.13.10:

Configured by nick at Sun Mar 13 15:23:34 GMT 2011.

Summary of my perl5 (revision 5 version 13 subversion 10) configuration:
  Derived from: cb07477ea9af3a143e6227a4db7952cd7b78dd76
  Platform:
    osname=linux, osvers=2.6.35.4, archname=x86_64-linux-thread-multi-ld
    uname='linux eris 2.6.35.4 #4 smp tue sep 21 09:54:22 bst 2010 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-Os -Dusethreads -Duselongdouble -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.13.10-427-gcb07477 -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.13.10-427-gcb07477 -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -Umad -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-Os',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='long double', nvsize=16, Off_t='off_t', lseeksize=8
    alignbytes=16, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -Os -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.13.10:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.13.10-427-gcb07477/lib/perl5/site_perl/5.13.10/x86_64-linux-thread-multi-ld
    /home/nick/Sandpit/snap5.9.x-v5.13.10-427-gcb07477/lib/perl5/site_perl/5.13.10
    /home/nick/Sandpit/snap5.9.x-v5.13.10-427-gcb07477/lib/perl5/5.13.10/x86_64-linux-thread-multi-ld
    /home/nick/Sandpit/snap5.9.x-v5.13.10-427-gcb07477/lib/perl5/5.13.10
    .


Environment for perl 5.13.10:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Sun Mar 13 09&#8203;:26&#8203;:57 2011\, nicholas wrote&#8203;:

> t/re/reg\_unsafe\.t was added in a229a030867b4d77\, because a match fails
>    when
> run alone\, but passes when in t/re/pat\.t \(subsequently moved to
>    pat\_advanced\.t\)

\[skip hate about folding and then quoting\]

Of course\, I've just helped "confuse" things a bit further by reuniting
the two tests\, by moving the meat of reg\_unsafe\.t into pat\_advanced\.t\,
thanks to the wonders of fresh\_perl\_is\(\)\.

http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/1b474ee3fedd2cefdaafdda6e060b8036ca756df

Nicholas Clark

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

I fixed this once\, and then we had to rip it out because the performance
implications were dire\.

Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they cause
all strings to be copied before matches \(to avoid the same "bug" this
code triggers\)\.

However this also causes the slowdown we wanted to avoid by not fixing it\.

The problem is code like this&#8203;:

my $str="something really big";
while \($str=~/\(\\w\+\)/g\) \{

\}

With the fix I did\, and with $& mentioned the above will copy $str each
time through\, and thus when $str is 1GB will become punishingly slow\.

I don't think this has an easy fix\.

cheers\,
Yves

**Comment by p5pRT** [comment]
@demerphq - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @demerphq

On 13 March 2011 20&#8203;:26\, yves orton via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> I fixed this once\, and then we had to rip it out because the performance
> implications were dire\.
>
> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they cause
> all strings to be copied before matches \(to avoid the same "bug" this
> code triggers\)\.
>
> However this also causes the slowdown we wanted to avoid by not fixing it\.
>
> The problem is code like this&#8203;:
>
> my $str="something really big";
> while \($str=~/\(\\w\+\)/g\) \{
>
> \}
>
> With the fix I did\, and with $& mentioned the above will copy $str each
> time through\, and thus when $str is 1GB will become punishingly slow\.
>
> I don't think this has an easy fix\.

And for the observant\, this case is actually special cased to NOT copy
the string even if capture buffers are used\.

Yves
\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Mar 13 12&#8203;:26&#8203;:26 2011\, demerphq wrote&#8203;:
> I fixed this once\, and then we had to rip it out because the performance
> implications were dire\.
> 
> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they cause
> all strings to be copied before matches \(to avoid the same "bug" this
> code triggers\)\.
> 
> However this also causes the slowdown we wanted to avoid by not fixing it\.
> 
> The problem is code like this&#8203;:
> 
> my $str="something really big";
> while \($str=~/\(\\w\+\)/g\) \{
> 
> \}
> 
> With the fix I did\, and with $& mentioned the above will copy $str each
> time through\, and thus when $str is 1GB will become punishingly slow\.
> 
> I don't think this has an easy fix\.

How about some variation of COW? \(I\.e\.\, a post\- \(as opposed to pre\-\)
match copy when the target is assigned to\.\)

**Comment by p5pRT** [comment]
### From @demerphq

On 13 March 2011 22&#8203;:47\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Sun Mar 13 12&#8203;:26&#8203;:26 2011\, demerphq wrote&#8203;:
>> I fixed this once\, and then we had to rip it out because the performance
>> implications were dire\.
>>
>> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they cause
>> all strings to be copied before matches \(to avoid the same "bug" this
>> code triggers\)\.
>>
>> However this also causes the slowdown we wanted to avoid by not fixing it\.
>>
>> The problem is code like this&#8203;:
>>
>> my $str="something really big";
>> while \($str=~/\(\\w\+\)/g\) \{
>>
>> \}
>>
>> With the fix I did\, and with $& mentioned the above will copy $str each
>> time through\, and thus when $str is 1GB will become punishingly slow\.
>>
>> I don't think this has an easy fix\.
>
> How about some variation of COW? \(I\.e\.\, a post\- \(as opposed to pre\-\)
> match copy when the target is assigned to\.\)

If its doable maybe\. But i have a feeling its a can of worms\.

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Mar 14 05&#8203;:01&#8203;:20 2011\, demerphq wrote&#8203;:
> On 13 March 2011 22&#8203;:47\, Father Chrysostomos via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > On Sun Mar 13 12&#8203;:26&#8203;:26 2011\, demerphq wrote&#8203;:
> >> I fixed this once\, and then we had to rip it out because the
> performance
> >> implications were dire\.
> >>
> >> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they
> cause
> >> all strings to be copied before matches \(to avoid the same "bug"
> this
> >> code triggers\)\.
> >>
> >> However this also causes the slowdown we wanted to avoid by not
> fixing it\.
> >>
> >> The problem is code like this&#8203;:
> >>
> >> my $str="something really big";
> >> while \($str=~/\(\\w\+\)/g\) \{
> >>
> >> \}
> >>
> >> With the fix I did\, and with $& mentioned the above will copy $str
> each
> >> time through\, and thus when $str is 1GB will become punishingly
> slow\.
> >>
> >> I don't think this has an easy fix\.
> >
> > How about some variation of COW? \(I\.e\.\, a post\- \(as opposed to pre\-\)
> > match copy when the target is assigned to\.\)
> 
> If its doable maybe\. But i have a feeling its a can of worms\.

Can someone point me to the code that does the pre\-match copy and the
code that sets up $1\, $2\, etc\.?

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 26 23&#8203;:32&#8203;:32 2011\, sprout wrote&#8203;:
> On Mon Mar 14 05&#8203;:01&#8203;:20 2011\, demerphq wrote&#8203;:
> > On 13 March 2011 22&#8203;:47\, Father Chrysostomos via RT
> > \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > > On Sun Mar 13 12&#8203;:26&#8203;:26 2011\, demerphq wrote&#8203;:
> > >> I fixed this once\, and then we had to rip it out because the
> > performance
> > >> implications were dire\.
> > >>
> > >> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they
> > cause
> > >> all strings to be copied before matches \(to avoid the same "bug"
> > this
> > >> code triggers\)\.
> > >>
> > >> However this also causes the slowdown we wanted to avoid by not
> > fixing it\.
> > >>
> > >> The problem is code like this&#8203;:
> > >>
> > >> my $str="something really big";
> > >> while \($str=~/\(\\w\+\)/g\) \{
> > >>
> > >> \}
> > >>
> > >> With the fix I did\, and with $& mentioned the above will copy $str
> > each
> > >> time through\, and thus when $str is 1GB will become punishingly
> > slow\.
> > >>
> > >> I don't think this has an easy fix\.
> > >
> > > How about some variation of COW? \(I\.e\.\, a post\- \(as opposed to pre\-\)
> > > match copy when the target is assigned to\.\)
> > 
> > If its doable maybe\. But i have a feeling its a can of worms\.
> 
> Can someone point me to the code that does the pre\-match copy and the
> code that sets up $1\, $2\, etc\.?

And by the latter I don’t mean gv\_fetchpvn\_flags\, but whatever sets up
the information that those access\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Mar 13 12&#8203;:59&#8203;:21 2011\, demerphq wrote&#8203;:
> On 13 March 2011 20&#8203;:26\, yves orton via RT \<perlbug\-followup@&#8203;perl\.org>
wrote&#8203;:
> > I fixed this once\, and then we had to rip it out because the performance
> > implications were dire\.
> >
> > Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they cause
> > all strings to be copied before matches \(to avoid the same "bug" this
> > code triggers\)\.
> >
> > However this also causes the slowdown we wanted to avoid by not
fixing it\.
> >
> > The problem is code like this&#8203;:
> >
> > my $str="something really big";
> > while \($str=~/\(\\w\+\)/g\) \{
> >
> > \}
> >
> > With the fix I did\, and with $& mentioned the above will copy $str each
> > time through\, and thus when $str is 1GB will become punishingly slow\.
> >
> > I don't think this has an easy fix\.
> 
> And for the observant\, this case is actually special cased to NOT copy
> the string even if capture buffers are used\.

David Mitchell \(git says it was David that time\, not Dave\) fixed this in
the branched merged to blead as 1703c1fc996f9\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Mar 14 05&#8203;:01&#8203;:20 2011\, demerphq wrote&#8203;:
> On 13 March 2011 22&#8203;:47\, Father Chrysostomos via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > On Sun Mar 13 12&#8203;:26&#8203;:26 2011\, demerphq wrote&#8203;:
> >> I fixed this once\, and then we had to rip it out because the
> performance
> >> implications were dire\.
> >>
> >> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they
> cause
> >> all strings to be copied before matches \(to avoid the same "bug"
> this
> >> code triggers\)\.
> >>
> >> However this also causes the slowdown we wanted to avoid by not
> fixing it\.
> >>
> >> The problem is code like this&#8203;:
> >>
> >> my $str="something really big";
> >> while \($str=~/\(\\w\+\)/g\) \{
> >>
> >> \}
> >>
> >> With the fix I did\, and with $& mentioned the above will copy $str
> each
> >> time through\, and thus when $str is 1GB will become punishingly
> slow\.
> >>
> >> I don't think this has an easy fix\.
> >
> > How about some variation of COW? \(I\.e\.\, a post\- \(as opposed to pre\-\)
> > match copy when the target is assigned to\.\)
> 
> If its doable maybe\. But i have a feeling its a can of worms\.

This can of worms is about to be merged\, once the smokers have had a go
at smoke\-me/ookow\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Nov 25 14&#8203;:16&#8203;:03 2012\, sprout wrote&#8203;:
> On Mon Mar 14 05&#8203;:01&#8203;:20 2011\, demerphq wrote&#8203;:
> > On 13 March 2011 22&#8203;:47\, Father Chrysostomos via RT
> > \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > > On Sun Mar 13 12&#8203;:26&#8203;:26 2011\, demerphq wrote&#8203;:
> > >> I fixed this once\, and then we had to rip it out because the
> > performance
> > >> implications were dire\.
> > >>
> > >> Mentioning $\` $& $' \*anywhere\* will "fix" the bug\, because they
> > cause
> > >> all strings to be copied before matches \(to avoid the same "bug"
> > this
> > >> code triggers\)\.
> > >>
> > >> However this also causes the slowdown we wanted to avoid by not
> > fixing it\.
> > >>
> > >> The problem is code like this&#8203;:
> > >>
> > >> my $str="something really big";
> > >> while \($str=~/\(\\w\+\)/g\) \{
> > >>
> > >> \}
> > >>
> > >> With the fix I did\, and with $& mentioned the above will copy $str
> > each
> > >> time through\, and thus when $str is 1GB will become punishingly
> > slow\.
> > >>
> > >> I don't think this has an easy fix\.
> > >
> > > How about some variation of COW? \(I\.e\.\, a post\- \(as opposed to pre\-\)
> > > match copy when the target is assigned to\.\)
> > 
> > If its doable maybe\. But i have a feeling its a can of worms\.
> 
> This can of worms is about to be merged\, once the smokers have had a go
> at smoke\-me/ookow\.

And it is now merged as cd298ce42eb\, fixing \#4289\.

\-\- 

Father Chrysostomos
